File: frontend/src/utils/buildInfo.ts
import type { BuildInfo } from '@/utils/api/auth'

export function formatBuildVersion(info: BuildInfo | null | undefined): string {
  if (!info?.version) return ''
  const commit = shortCommitSha(info.build?.commit_sha)
  return commit ? `Core v${info.version} - ${commit}` : `Core v${info.version}`
}

export function buildInfoTitle(info: BuildInfo | null | undefined): string {
  if (!info) return ''
  const parts = [
    `${info.name} v${info.version}`,
    `API ${info.api_version}`,
    info.build?.image_tag ? `image ${info.build.image_tag}` : null,
    info.build?.commit_sha ? `commit ${info.build.commit_sha}` : null,
    info.build?.built_at ? `built ${info.build.built_at}` : null,
  ].filter((part): part is string => Boolean(part))
  return parts.join(' - ')
}

function shortCommitSha(commitSha: string | null | undefined): string {
  const normalized = commitSha?.trim()
  if (!normalized) return ''
  return normalized.slice(0, 7)
}

File: src/kokoro_link/api/contracts/build_info.py
from __future__ import annotations

from pydantic import BaseModel

from kokoro_link.infrastructure.build_info import BuildInfo


class BuildMetadataResponse(BaseModel):
    image_tag: str | None = None
    commit_sha: str | None = None
    built_at: str | None = None


class BuildInfoResponse(BaseModel):
    name: str
    version: str
    api_version: str
    build: BuildMetadataResponse


def build_info_response(info: BuildInfo) -> BuildInfoResponse:
    return BuildInfoResponse(
        name=info.name,
        version=info.version,
        api_version=info.api_version,
        build=BuildMetadataResponse(
            image_tag=info.build.image_tag,
            commit_sha=info.build.commit_sha,
            built_at=info.build.built_at,
        ),
    )
